Passenger 5 support.

pjkelly opened the pull request:
For your consideration.  In addition to Passenger 5 support, this PR also includes some (opinionated) refactoring of the plugin with the goal of making it a little more flexible.  I'm absolutely open to discussion and/or revisions on implementation here.
- Added support for Passenger 5's new restart mechanism and made it the default.
- Added new configuration options.
- Changed namespace from `:deploy` to `:passenger` since `:deploy` is commonly defined/used in users' deploy.rb files.
- Removed automatic placement of `passenger:restart` in the `after :publishing` callback. This results in the plugin being less plug-and-play, but more flexible i.e. users can call `passenger:restart` wherever and whenever they want in their deployment flow.
- Updated README based on all the above changes.

This PR addresses #2.

jalada commented:
Why not use `deploy_to` instead of `deploy_path`, which just wraps the `deploy_to` string in a Pathname object?

jalada commented:
I think it's also worthwhile enabling Passenger's `--ignore-app-not-running` option by default, because Capistrano halts the deploy if the app isn't running already, which doesn't make a lot of sense (especially on the first deploy, for example).

betesh commented:
@pjkelly I have a few concerns with this PR, though I certainly appreciate the effort.

1) Thank you for bringing to my attention that the URL in the README was out-of-date.  I've corrected it in another commit.

2) Why does the user need to tell capistrano what version of passenger is running?  We should be detecting that automatically on a server-by-server basis.  For your reference, here's the output when I check the version on a machine that has v4 installed:

```
[myUser@myServer]$ passenger -v
Phusion Passenger version 4.0.57

"Phusion Passenger" is a trademark of Hongli Lai & Ninh Bui.
[myUser@myServer]$  passenger-config --version
4.0.57
```

Can you use one of these commands to check the version instead of requiring the user to specify it?

3) Changing the namespace is a breaking change.  While I understand your reasons for it, if you want to make the change, we need to support both and log a deprecation warning for now.  If you would like to do this, please create a separate PR for it.

pjkelly commented:
@betesh & @jalada - awesome feedback - thanks for getting back to me so quickly.

I'm with you on all this.  I think the best course of action is to adjust this PR so that it only focuses on the Passenger 5 support, and then I'll open a separate PR with the namespace stuff - in retrospect, I should've done that in the first place.

I'll get these revisions going soon.

betesh commented:
@pjkelly Thanks, looking forward to it!

pjkelly committed:
Passenger 5 support.

This is a backwards-compatible implementation that checks the version of Passenger installed on the server and executes the restart command based on that information.

pjkelly commented:
@betesh - I just pushed the changes we discussed for you to review.  Let me know if you have any questions/feedback.

betesh commented:
I still have 1 strong objection: You're making the assumption that all servers are running the same version of passenger.  This is what I was referring to in my previous comment about "detecting that automatically **on a server-by-server basis**"

If instead of 3 separate tasks, you run all the logic in a single on/with block, it will reduce code repetition (making the code more maintainable--I have very strong opinions about unnecessary duplication and its impact on maintainability), and it will also allow each server to run a different version of passenger.  Lastly, supporting 2 separate restart tasks that can be called in isolation gives the impression that it would make sense to call either one.

Also, wouldn't it be easier to use `passenger-config --version` than `passenger -v`, since it only outputs the version number?  Does `passenger-config --version` work in Passenger 5?

jalada commented:
I can confirm it works with 5:

```
$ passenger-config --version
Phusion Passenger 5.0.4
```

pjkelly commented:
@betesh - thanks for the feedback - the duplication was bugging me too.  I thought the way it was setup would support multiple versions of passenger at the same time, but was mistaken.  I went ahead boiled everything down into a single task now (`passenger:restart`) that works the way we’ve discussed.  I ran the plugin against two servers with different versions of Passenger and they were both restarted with the appropriate commands.

**Version Checking**

The reason I elected to go with `passenger -v` is because the output format of `passenger-config --version` changed at some point along the way:

``` bash
# Passenger 3
$ passenger-config --version
3.0.9
$ passenger -v
Phusion Passenger version 3.0.9

“Phusion Passenger” is a trademark of Hongli Lai & Ninh Bui.

# Passenger 5
$ passenger-config --version
Phusion Passenger 5.0.4
$ passenger -v
Phusion Passenger version 5.0.4

“Phusion Passenger” is a trademark of Hongli Lai & Ninh Bui.
```

I don’t have a server running Passenger 4, so I couldn’t verify that at the moment, but this discrepancy was enough to suggest that going with `passenger -v` was the simpler option.

betesh commented:
Looks good.  I will test it against passenger 4 when I get a chance and if all goes well, I will merge it.

pjkelly commented:
@betesh - :+1:

betesh commented:
@pjkelly I added a task to add passenger and passenger-config to the list of commands that need to be prefixed by `rvm`.  This made `passenger -v` work in passenger 4 with no additional configuration.

Can you remove whatever overridden values you are using for `passenger_environment_variables` and test that this works in passenger 3 & 5 (assuming you are using RVM)?

pjkelly commented:
@betesh - we don't use RVM, so I don't have the ability to test that - apologies.

betesh commented:
@pjkelly In that case can you take a quick look at the commits I just added to make sure they look sane enough and let me know if you spot any obvious mistakes?  Link for easy comparing: https://github.com/crushlovely/passenger/compare/passenger-5-support-and-other-adjustments...capistrano:master

pjkelly commented:
It looks good to me!

jalada commented:
I'll confirm on Monday but Passenger 5 generally installs outside of RVM, even if you have RVM:

```
$ which rvm
/usr/local/rvm/bin/rvm
$ which passenger
/usr/bin/passenger
```

This is because Passenger 5 is C++ and language agnostic now. It might not cause a problem, but like I say, I'll double check :)

betesh commented:
Thanks @jalada I was unaware.  Please let me know once you confirm.

betesh commented:
@jalada I installed passenger 5 on one of our servers this morning, and it is thoroughly installed through RVM

```
[my_user@my_host ~]$ which rvm
~/.rvm/bin/rvm
[my_user@my_host ~]$ which passenger
~/.rvm/gems/ruby-2.1.3/bin/passenger
[my_user@my_host ~]$ passenger -v
Phusion Passenger version 5.0.5

"Phusion Passenger" is a trademark of Hongli Lai & Ninh Bui.
[my_user@my_host ~]$ passenger-config --version
Phusion Passenger 5.0.5
```

In your case, you'll probably need to override the passenger:rvm:hook task so that passenger does not end up appended to `rvm_map_bins`.

jalada commented:
@betesh yes but that isn't necessarily required nor recommended. Passenger is now language agnostic and the version of Ruby you run Passenger as is not dependent on the version of Ruby you run your app with. In fact you can use multiple versions of Ruby on the same installation of Passenger. Therefore automatically tying Passenger to the RVM you are deploying your app with doesn't make sense; that should be an opt in, not an opt out.

[More information about Passenger 5 and the relation with Ruby](https://www.phusionpassenger.com/documentation/Users%20guide%20Apache.html#relationship_with_ruby)

betesh commented:
@jalada Thank you for the link.  I took a look and I think the best approach would be to see if there's a way we could automatically detect where passenger is installed and decide based on that whether to add it to the command map.  No opting (in or out) should be necessary.

Suppose we run `which passenger` and see what the result is.  If it's not found in any $PATH directories, and RVM is being used, than I think it's reasonable to assume that it's installed through RVM.  If it is found, of course, than we don't add it to the command map.

Thoughts?  Opinions?

jalada commented:
Yeah, if that's possible it sounds like a plan. It's a shame that Capistrano RVM works like that and doesn't just `source` in the environment in the shell it's running, because then `passenger` would be resolved based on the `PATH` including RVM paths if they are included. Is there no way of doing that? (I'm unfamiliar with the workings of `capistrano-rvm`)

betesh commented:
@jalada Try it with the lastest commit (68b86f1) and let me know if this works for you out-of-the-box.
